gpt-survey-summarizer/templates/deployment.yaml

```diff
@@ -0,0 +1,65 @@
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: {{ include "gpt-survey-summarizer.fullname" . }}
+  labels:
+    {{- include "gpt-survey-summarizer.labels" . | nindent 4 }}
+spec:
+  {{- if not .Values.autoscaling.enabled }}
+  replicas: {{ .Values.replicaCount }}
+  {{- end }}
+  selector:
+    matchLabels:
+      {{- include "gpt-survey-summarizer.selectorLabels" . | nindent 6 }}
+  template:
+    metadata:
+      {{- with .Values.podAnnotations }}
+      annotations:
+        {{- toYaml . | nindent 8 }}
+      {{- end }}
+      labels:
+        {{- include "gpt-survey-summarizer.selectorLabels" . | nindent 8 }}
+    spec:
+      {{- with .Values.imagePullSecrets }}
+      imagePullSecrets:
+        {{- toYaml . | nindent 8 }}
+      {{- end }}
+      serviceAccountName: {{ include "gpt-survey-summarizer.serviceAccountName" . }}
+      securityContext:
+        {{- toYaml .Values.podSecurityContext | nindent 8 }}
+      containers:
+        - name: backend
+          securityContext:
+            {{- toYaml .Values.securityContext | nindent 12 }}
+          image: "{{ .Values.image.repository }}:{{ .Values.image.tag | default .Chart.AppVersion }}"
+          imagePullPolicy: {{ .Values.image.pullPolicy }}
+          env:
+            {{- if .Values.envVars }}
+            {{- toYaml .Values.envVars | nindent 12 }}
+            {{- end }}
```

Comment:
**suggestion:** I think it would be nice to abstract the process configuration here.

Instead of injecting the configuration like so:
```
envVars:
  - name: DISCORD_TOKEN
    value: token
  - name: OPENAI_API_TOKEN
    value: token
  - name: REDIS_URL
    value: url 
  - name: CLIENT_ID
    value: id
  - name: GUILD_ID
    value: id
  - name: SUMMARIZE_FREQUENCY_SECONDS
    value: "3600"
```

We could provide this interface:
```yaml
config:
  clientId: clientId
  discordToken: token
  guildId: guildId
  redisPassword: password
  summarizeFrequencySeconds: 3600
 ```

Then, this would allow to store the Discord token, OpenAI token and Redis password/url in a Kubernetes `Secret`, to be loaded by the pod and injected as envvar using, e.g.:
```
...
       env:
       - name: DISCORD_TOKEN
         valueFrom:
           secretKeyRef:
             name: bot-secrets
             key: discordToken
...
```

Another reason is it is a [good practice to store sensitive in Secrets](https://kubernetes.io/docs/concepts/security/secrets-good-practices).

Reply:
For the context of this comment the changes in this [PR](https://github.com/MinaFoundation/gitops-staging/pull/160/files) are relevant

1. I agree on having `bot.envVars` and `summarizer.envVars` now that we have two operation modes and two different deployments, consider it done.
2. Each implementation method of environment variables has its benefits and drawbacks:

- **pros of current implementation:** we don't have to change the helm chart when adding a new variable and there is no abstraction between the value file and the variables in the container (values and variable names match 1:1) **cons:** we cannot define default values for the environment variables
- **pros of suggested implementation:** we can define default values in `values.yaml` and only declare specific env vars **cons:** each time a new environment variable is added we have to implement it in the chart, `values.yaml` and `helmfile.yaml`

We can hold a vote which one we stick to.

3. On the topic of `Secret` management, [helmfile manages secrets](https://helmfile.readthedocs.io/en/stable/remote-secrets/) with helm-secrets plugin, which at the end uses Kubernetes Secrets (if I am not mistaken, documentation is not too detailed about this, and the source code didn't really help).

Currently after deployment there is a secret holding the fetched values from AWS SM

```
    ❯ k get secret
    NAME                          TYPE     DATA   AGE
    gpt-survey-summarizer-redis   Opaque   1      2d7h
```

So with this in mind
```
      envVars:
        - name: DISCORD_TOKEN
          value: {{ .Values.secrets.discord_token | fetchSecretValue }}
```

Could yield the desired results.

Reply:
I am happy to keep the current implementation given the early stage of the project.

That being said, regarding:
> cons: each time a new environment variable is added we have to implement it in the chart, values.yaml and helmfile.yaml

Yes, it's a valid concern. This is usually done using a mix of too approaches:
- a `config` object to define the abstraction to configure the process.
- an `extraEnvVars` to define extra environment variables.

Note that the word "extra" is important. It shows that the  `config` and `extraEnvVars` does not overwrite each other, therefore it is possible to define defaults.

See for example:
- in the `values.yaml`: https://github.com/bitnami/charts/blob/8c566496419450b8dec3e908b563b2dbacbb149c/bitnami/apache/values.yaml#L372-L374
- in the `template`: https://github.com/bitnami/charts/blob/8c566496419450b8dec3e908b563b2dbacbb149c/bitnami/apache/templates/deployment.yaml#L176-L185

Regarding:
> On the topic of Secret management, [helmfile manages secrets](https://helmfile.readthedocs.io/en/stable/remote-secrets/) with helm-secrets plugin, which at the end uses Kubernetes Secrets (if I am not mistaken, documentation is not too detailed about this, and the source code didn't really help).

There is two subject here: templating and manifests.

For the templating, we are using both `helmfile` and `helm`. When rendering sensitive, we need to make sure they are fetch from a secure source, and injected in the K8s manifests. The `helmfile` documentation specifies that sensitives can be fetch from remote vaults, or using `helm-secrets` plugin (which stores sensitives on the local repository using SOPS).

My suggestion was about the second part, for the manifests. It is recommended that K8s manifests, once rendered, should declare sensitive in K8s `Secret` objects. Then, K8s `Deployment` objects would load them at runtime.

As I said, I am happy to keep the current implementation given the early stage of the project.

Reply:
I understand both of your points.

Obviously the second one empowers the first one so there is not much to debate about. I like the approach of `config` + `extraEnvVars` because it is a best of both worlds approach. Implement variables that have specific implementation needs and inject everything else.

I will implement these, thanks for explaining these things to me.